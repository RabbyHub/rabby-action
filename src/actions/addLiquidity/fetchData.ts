import PQueue from 'p-queue';
import {
  AddLiquidityRequireData,
  FetchActionRequiredData,
  ParsedActionData,
} from '../../types';
import { waitQueueFinished } from '../../utils/waitQueueFinished';
import BigNumber from 'bignumber.js';

export const fetchDataAddLiquidity: FetchActionRequiredData<{
  receiver: string;
  exchange_rate: string;
  min_exchange_rate: string;
  max_exchange_rate: string;
  token0: any;
  token1: any;
}> = async (options, likeAction) => {
  const { actionData, apiProvider, sender, walletProvider, chainId } = options;

  const action =
    likeAction || (<ParsedActionData<'transaction'>>actionData).addLiquidity;

  if (!action || options.type === 'text') {
    return {};
  }

  const { receiver, exchange_rate, token0, token1 } = action;

  const fair_exchange_rate = new BigNumber(token0.price).div(token1.price);

  const diff = +new BigNumber(exchange_rate)
    .minus(fair_exchange_rate)
    .abs()
    .div(fair_exchange_rate)
    .times(100)
    .toFixed(4);

  const receiverInWallet = await walletProvider.hasAddress(receiver);

  const id =
    options.type === 'transaction'
      ? options.tx.to
      : options.actionData.contractId;
  if (!id) {
    return {};
  }

  const result: AddLiquidityRequireData = {
    id,
    protocol: null,
    bornAt: 0,
    rank: null,
    sender,
    receiverInWallet,
    hasInteraction: false,
    receiver,
    poolRate: exchange_rate,
    token0,
    token1,
    marketRate: fair_exchange_rate.toFixed(),
    diff,
  };
  const queue = new PQueue();

  queue.add(async () => {
    const contractInfo = await apiProvider.getContractInfo(id, chainId);
    if (!contractInfo) {
      result.rank = null;
    } else {
      result.rank = contractInfo.credit.rank_at;
      result.bornAt = contractInfo.create_at;

      result.protocol = contractInfo.protocol;
    }
  });

  queue.add(async () => {
    const hasInteraction = await apiProvider.hasInteraction(
      sender,
      chainId,
      id
    );
    result.hasInteraction = hasInteraction.has_interaction;
  });

  await waitQueueFinished(queue);

  return result;
};
