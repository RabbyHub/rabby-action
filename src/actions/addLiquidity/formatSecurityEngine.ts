import {
  AddLiquidityRequireData,
  FormatSecurityEngineContext,
} from '../../types';

export const formatSecurityEngineAddLiquidity: FormatSecurityEngineContext =
  async (options) => {
    const { actionData, requireData, chainId } = options;

    if (!actionData.addLiquidity || !chainId) {
      return {} as any;
    }

    const data = requireData as AddLiquidityRequireData;

    return {
      addLiquidity: {
        receiverMatch:
          data.receiver?.toLowerCase() === data.sender.toLowerCase(),
        diff: data.diff,
      },
    };
  };
