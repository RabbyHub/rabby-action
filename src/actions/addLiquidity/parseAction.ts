import { ParseAction } from '../../types';
import { TokenItem } from '@rabby-wallet/rabby-api/dist/types';

interface AddLiquidityActionData {
  receiver: string;
  exchange_rate: string;
  min_exchange_rate: string;
  max_exchange_rate: string;
  token0: TokenItem;
  token1: TokenItem;
}

export const parseActionAddLiquidity: ParseAction = (options) => {
  const { data } = options;

  if (data?.type !== 'add_liquidity') {
    return {};
  }

  const {
    receiver,
    exchange_rate,
    min_exchange_rate,
    max_exchange_rate,
    token0,
    token1,
  } = data.data as unknown as AddLiquidityActionData;

  return {
    addLiquidity: {
      receiver,
      exchange_rate,
      min_exchange_rate,
      max_exchange_rate,
      token0,
      token1,
    },
  };
};
