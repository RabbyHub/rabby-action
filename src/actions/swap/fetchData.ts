import PQueue from 'p-queue';
import { waitQueueFinished } from '../../utils/waitQueueFinished';
import { FetchActionRequiredData, SwapRequireData } from '../../types';
import { fetchHasInteraction } from '../../fetches/fetchHasInteraction';

// TODO: text
export const fetchDataSwap: FetchActionRequiredData<{
  receiver: string;
}> = async (options, likeSwapAction) => {
  if (options.type === 'text') {
    return {};
  }
  const queue = new PQueue();
  const {
    actionData,
    walletProvider,
    sender,
    apiProvider,
    chainId,
    extraActionDataState,
  } = options;
  const swapAction = likeSwapAction || actionData.swap;
  if (!swapAction) {
    return {};
  }

  const { receiver } = swapAction;

  const receiverInWallet = await walletProvider.hasAddress(receiver);
  const id =
    options.type === 'transaction'
      ? options.tx.to
      : options.actionData.contractId;

  if (!id) {
    return {};
  }

  const result: SwapRequireData = {
    id,
    protocol: null,
    bornAt: 0,
    rank: null,
    sender,
    receiverInWallet,
    hasInteraction: null,
    extraState: {
      hasInteraction: () => undefined,
    },
  };
  queue.add(async () => {
    const contractInfo = await apiProvider.getContractInfo(id, chainId);
    if (!contractInfo) {
      result.rank = null;
    } else {
      result.rank = contractInfo.credit.rank_at;
      result.bornAt = contractInfo.create_at;
      result.protocol = contractInfo.protocol;
    }
  });

  fetchHasInteraction({
    apiProvider,
    sender,
    chainId,
    spender: id,
    extraActionDataState,
    queue,
    result,
  });

  await waitQueueFinished(queue);
  return result;
};
