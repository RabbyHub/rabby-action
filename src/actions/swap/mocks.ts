export const preExecData = {
  abi: {
    func: 'swap',
    params: [
      '0x1E82AD8A12068A85fCb96368463B434e77b21201',
      [
        '0x0000000000000000000000000000000000001010',
        '0xc2132D05D31c914a87C6611C10748AEb04B58e8F',
        '0x1E82AD8A12068A85fCb96368463B434e77b21201',
        '0x5853eD4f26A3fceA565b3FBC698bb19cdF6DEB85',
        192647154008608,
        50,
        100,
        0,
        '0x39041F1B366fE33F9A5a79dE5120F2Aee2577ebc',
        '',
      ],
      [
        [
          0,
          0,
          0,
          'd1660f99000000000000000000000000000000000000000000000000000000000000101000000000000000000000000039041f1b366fe33f9a5a79de5120f2aee2577ebc0000000000000000000000000000000000000000000000000000007022a9b75d',
        ],
        [
          76001907929418866304487584807096610665432814192,
          0,
          192165536123587,
          'd0e30db0',
        ],
        [
          0,
          0,
          0,
          'eb5625d90000000000000000000000000d500b1d8e8ef31e21c99d1db9a6444d3adf1270000000000000000000000000ba12222222228d8ba445958a75a0704d566bf2c80000000000000000000000000000000000000000000000000000aec60684cac3',
        ],
        [
          1062276670219503018717391839032010498165418226376,
          0,
          0,
          '52bbbe2900000000000000000000000000000000000000000000000000000000000000e00000000000000000000000001e82ad8a12068a85fcb96368463b434e77b2120100000000000000000000000000000000000000000000000000000000000000000000000000000000000000001e82ad8a12068a85fcb96368463b434e77b212010000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000ffffffff5e6989c0e2b6600ab585d56bf05479d5450a60c800010000000000000000005500000000000000000000000000000000000000000000000000000000000000000000000000000000000000000d500b1d8e8ef31e21c99d1db9a6444d3adf1270000000000000000000000000c2132d05d31c914a87c6611c10748aeb04b58e8f0000000000000000000000000000000000000000000000000000aec60684cac300000000000000000000000000000000000000000000000000000000000000c00000000000000000000000000000000000000000000000000000000000000000',
        ],
        [
          0,
          0,
          0,
          '8a6a1e85000000000000000000000000c2132d05d31c914a87c6611c10748aeb04b58e8f000000000000000000000000353c1f0bc78fbbc245b3c93ef77b1dcc5b77d2a00000000000000000000000000000000000000000000000000000000000000064',
        ],
        [
          0,
          0,
          0,
          '9f865422000000000000000000000000c2132d05d31c914a87c6611c10748aeb04b58e8f00000000000000000000000000000001000000000000000000000000000000010000000000000000000000000000000000000000000000000000000000000080000000000000000000000000000000000000000000000000000000000000004400000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000800000000000000000000000000000000000000000000000000000000000000064d1660f99000000000000000000000000c2132d05d31c914a87c6611c10748aeb04b58e8f0000000000000000000000005853ed4f26a3fcea565b3fbc698bb19cdf6deb85000000000000000000000000000000000000000000000000000000000000000100000000000000000000000000000000000000000000000000000000',
        ],
      ],
    ],
  },
  abi_str:
    '{\n    "func": "swap",\n    "params": [\n        "0x1E82AD8A12068A85fCb96368463B434e77b21201",\n        [\n            "0x0000000000000000000000000000000000001010",\n            "0xc2132D05D31c914a87C6611C10748AEb04B58e8F",\n            "0x1E82AD8A12068A85fCb96368463B434e77b21201",\n            "0x5853eD4f26A3fceA565b3FBC698bb19cdF6DEB85",\n            192647154008608,\n            50,\n            100,\n            0,\n            "0x39041F1B366fE33F9A5a79dE5120F2Aee2577ebc",\n            ""\n        ],\n        [\n            [\n                0,\n                0,\n                0,\n                "d1660f99000000000000000000000000000000000000000000000000000000000000101000000000000000000000000039041f1b366fe33f9a5a79de5120f2aee2577ebc0000000000000000000000000000000000000000000000000000007022a9b75d"\n            ],\n            [\n                76001907929418866304487584807096610665432814192,\n                0,\n                192165536123587,\n                "d0e30db0"\n            ],\n            [\n                0,\n                0,\n                0,\n                "eb5625d90000000000000000000000000d500b1d8e8ef31e21c99d1db9a6444d3adf1270000000000000000000000000ba12222222228d8ba445958a75a0704d566bf2c80000000000000000000000000000000000000000000000000000aec60684cac3"\n            ],\n            [\n                1062276670219503018717391839032010498165418226376,\n                0,\n                0,\n                "52bbbe2900000000000000000000000000000000000000000000000000000000000000e00000000000000000000000001e82ad8a12068a85fcb96368463b434e77b2120100000000000000000000000000000000000000000000000000000000000000000000000000000000000000001e82ad8a12068a85fcb96368463b434e77b212010000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000ffffffff5e6989c0e2b6600ab585d56bf05479d5450a60c800010000000000000000005500000000000000000000000000000000000000000000000000000000000000000000000000000000000000000d500b1d8e8ef31e21c99d1db9a6444d3adf1270000000000000000000000000c2132d05d31c914a87c6611c10748aeb04b58e8f0000000000000000000000000000000000000000000000000000aec60684cac300000000000000000000000000000000000000000000000000000000000000c00000000000000000000000000000000000000000000000000000000000000000"\n            ],\n            [\n                0,\n                0,\n                0,\n                "8a6a1e85000000000000000000000000c2132d05d31c914a87c6611c10748aeb04b58e8f000000000000000000000000353c1f0bc78fbbc245b3c93ef77b1dcc5b77d2a00000000000000000000000000000000000000000000000000000000000000064"\n            ],\n            [\n                0,\n                0,\n                0,\n                "9f865422000000000000000000000000c2132d05d31c914a87c6611c10748aeb04b58e8f00000000000000000000000000000001000000000000000000000000000000010000000000000000000000000000000000000000000000000000000000000080000000000000000000000000000000000000000000000000000000000000004400000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000800000000000000000000000000000000000000000000000000000000000000064d1660f99000000000000000000000000c2132d05d31c914a87c6611c10748aeb04b58e8f0000000000000000000000005853ed4f26a3fcea565b3fbc698bb19cdf6deb85000000000000000000000000000000000000000000000000000000000000000100000000000000000000000000000000000000000000000000000000"\n            ]\n        ]\n    ]\n}',
  balance_change: {
    success: true,
    error: null,
    send_token_list: [
      {
        id: 'matic',
        chain: 'matic',
        name: 'MATIC',
        symbol: 'MATIC',
        display_symbol: null,
        optimized_symbol: 'MATIC',
        decimals: 18,
        logo_url:
          'https://static.debank.com/image/matic_token/logo_url/matic/6f5a6b6f0732a7a235131bd7804d357c.png',
        protocol_id: '',
        price: 0.5226,
        price_24h_change: -0.08716157205240184,
        credit_score: 100000000.0,
        is_verified: true,
        is_scam: false,
        is_suspicious: false,
        is_core: true,
        is_wallet: true,
        time_at: null,
        amount: 0.000192647154008608,
        usd_value: 0.00010067740268489854,
        raw_amount: '192647154008608',
      },
    ],
    receive_token_list: [
      {
        id: '0xc2132d05d31c914a87c6611c10748aeb04b58e8f',
        chain: 'matic',
        name: '(PoS) Tether USD',
        symbol: 'USDT',
        display_symbol: null,
        optimized_symbol: 'USDT',
        decimals: 6,
        logo_url:
          'https://static.debank.com/image/eth_token/logo_url/0xdac17f958d2ee523a2206206994597c13d831ec7/464c0de678334b8fe87327e527bc476d.png',
        protocol_id: '',
        price: 1.00038,
        price_24h_change: 9.996301368559145e-6,
        credit_score: 4530369.670714375,
        is_verified: true,
        is_scam: false,
        is_suspicious: false,
        is_core: true,
        is_wallet: true,
        time_at: 1599512847.0,
        amount: 0.0001,
        usd_value: 0.00010003800000000001,
        raw_amount: '100',
      },
    ],
    send_nft_list: [],
    receive_nft_list: [],
    usd_value_change: -6.394026848985336e-7,
  },
  gas: {
    success: true,
    error: null,
    gas_used: 231360,
    gas_limit: 0,
    gas_ratio: 1.3,
  },
  is_gnosis: false,
  native_token: {
    id: 'matic',
    chain: 'matic',
    name: 'MATIC',
    symbol: 'MATIC',
    display_symbol: null,
    optimized_symbol: 'MATIC',
    decimals: 18,
    logo_url:
      'https://static.debank.com/image/matic_token/logo_url/matic/6f5a6b6f0732a7a235131bd7804d357c.png',
    protocol_id: '',
    price: 0.5226,
    price_24h_change: -0.08716157205240184,
    credit_score: 100000000.0,
    is_verified: true,
    is_scam: false,
    is_suspicious: false,
    is_core: true,
    is_wallet: true,
    time_at: null,
    amount: 0.000192647154008608,
  },
  pre_exec_version: 'v2',
  pre_exec: {
    success: true,
    error: null,
  },
  type_call: {
    action: 'swap',
    contract: '0x6352a56caadc4f1e25cd6c75970fa768a3304e64',
    contract_protocol_name: 'OpenOcean',
    contract_protocol_logo_url:
      'https://static.debank.com/image/project/logo_url/matic_openocean/b4fcab0bb9ebc86e8ecebf45b381b8a5.png',
  },
  trace_id: 'a0780bc9f16644718dd010d5fdec2254',
} as any;

export const parseTxData = {
  action: {
    type: 'swap_token',
    data: {
      pay_token: {
        id: 'matic',
        chain: 'matic',
        name: 'MATIC',
        symbol: 'MATIC',
        display_symbol: null,
        optimized_symbol: 'MATIC',
        decimals: 18,
        logo_url:
          'https://static.debank.com/image/matic_token/logo_url/matic/6f5a6b6f0732a7a235131bd7804d357c.png',
        protocol_id: '',
        price: 0.5226,
        price_24h_change: -0.08716157205240184,
        credit_score: 100000000.0,
        is_verified: true,
        is_scam: false,
        is_suspicious: false,
        is_core: true,
        is_wallet: true,
        time_at: null,
        amount: 0.000192647154008608,
        raw_amount: '192647154008608',
      },
      receive_token: {
        id: '0xc2132d05d31c914a87c6611c10748aeb04b58e8f',
        chain: 'matic',
        name: '(PoS) Tether USD',
        symbol: 'USDT',
        display_symbol: null,
        optimized_symbol: 'USDT',
        decimals: 6,
        logo_url:
          'https://static.debank.com/image/eth_token/logo_url/0xdac17f958d2ee523a2206206994597c13d831ec7/464c0de678334b8fe87327e527bc476d.png',
        protocol_id: '',
        price: 1.00038,
        price_24h_change: 9.996301368559145e-6,
        credit_score: 4530369.670714375,
        is_verified: true,
        is_scam: false,
        is_suspicious: false,
        is_core: true,
        is_wallet: true,
        time_at: 1599512847.0,
        min_amount: 5e-5,
        min_raw_amount: '50',
      },
      receiver: '0x5853ed4f26a3fcea565b3fbc698bb19cdf6deb85',
    },
    is_asset_changed: true,
    is_involving_privacy: false,
  },
  contract_call: {
    func: 'swap',
    contract: {
      id: '0x6352a56caadc4f1e25cd6c75970fa768a3304e64',
      protocol: {
        name: 'OpenOcean',
        logo_url:
          'https://static.debank.com/image/project/logo_url/matic_openocean/b4fcab0bb9ebc86e8ecebf45b381b8a5.png',
      },
    },
  },
  log_id: 87834234,
} as any;

export const txData = {
  chainId: 137,
  from: '0x5853ed4f26a3fcea565b3fbc698bb19cdf6deb85',
  to: '0x6352a56caadc4f1e25cd6c75970fa768a3304e64',
  value: '0xaf36292e8220',
  data: '0x90411a320000000000000000000000001e82ad8a12068a85fcb96368463b434e77b21201000000000000000000000000000000000000000000000000000000000000006000000000000000000000000000000000000000000000000000000000000001c00000000000000000000000000000000000000000000000000000000000001010000000000000000000000000c2132d05d31c914a87c6611c10748aeb04b58e8f0000000000000000000000001e82ad8a12068a85fcb96368463b434e77b212010000000000000000000000005853ed4f26a3fcea565b3fbc698bb19cdf6deb850000000000000000000000000000000000000000000000000000af36292e822000000000000000000000000000000000000000000000000000000000000000320000000000000000000000000000000000000000000000000000000000000064000000000000000000000000000000000000000000000000000000000000000000000000000000000000000039041f1b366fe33f9a5a79de5120f2aee2577ebc00000000000000000000000000000000000000000000000000000000000001400000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000600000000000000000000000000000000000000000000000000000000000000c000000000000000000000000000000000000000000000000000000000000001e000000000000000000000000000000000000000000000000000000000000002a000000000000000000000000000000000000000000000000000000000000003c00000000000000000000000000000000000000000000000000000000000000640000000000000000000000000000000000000000000000000000000000000076000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000800000000000000000000000000000000000000000000000000000000000000064d1660f99000000000000000000000000000000000000000000000000000000000000101000000000000000000000000039041f1b366fe33f9a5a79de5120f2aee2577ebc0000000000000000000000000000000000000000000000000000007022a9b75d000000000000000000000000000000000000000000000000000000000000000000000000000000000d500b1d8e8ef31e21c99d1db9a6444d3adf127000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000aec60684cac300000000000000000000000000000000000000000000000000000000000000800000000000000000000000000000000000000000000000000000000000000004d0e30db00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000800000000000000000000000000000000000000000000000000000000000000064eb5625d90000000000000000000000000d500b1d8e8ef31e21c99d1db9a6444d3adf1270000000000000000000000000ba12222222228d8ba445958a75a0704d566bf2c80000000000000000000000000000000000000000000000000000aec60684cac300000000000000000000000000000000000000000000000000000000000000000000000000000000ba12222222228d8ba445958a75a0704d566bf2c800000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000008000000000000000000000000000000000000000000000000000000000000001c452bbbe2900000000000000000000000000000000000000000000000000000000000000e00000000000000000000000001e82ad8a12068a85fcb96368463b434e77b2120100000000000000000000000000000000000000000000000000000000000000000000000000000000000000001e82ad8a12068a85fcb96368463b434e77b212010000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000ffffffff5e6989c0e2b6600ab585d56bf05479d5450a60c800010000000000000000005500000000000000000000000000000000000000000000000000000000000000000000000000000000000000000d500b1d8e8ef31e21c99d1db9a6444d3adf1270000000000000000000000000c2132d05d31c914a87c6611c10748aeb04b58e8f0000000000000000000000000000000000000000000000000000aec60684cac300000000000000000000000000000000000000000000000000000000000000c0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000008000000000000000000000000000000000000000000000000000000000000000648a6a1e85000000000000000000000000c2132d05d31c914a87c6611c10748aeb04b58e8f000000000000000000000000353c1f0bc78fbbc245b3c93ef77b1dcc5b77d2a0000000000000000000000000000000000000000000000000000000000000006400000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000008000000000000000000000000000000000000000000000000000000000000001a49f865422000000000000000000000000c2132d05d31c914a87c6611c10748aeb04b58e8f00000000000000000000000000000001000000000000000000000000000000010000000000000000000000000000000000000000000000000000000000000080000000000000000000000000000000000000000000000000000000000000004400000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000800000000000000000000000000000000000000000000000000000000000000064d1660f99000000000000000000000000c2132d05d31c914a87c6611c10748aeb04b58e8f0000000000000000000000005853ed4f26a3fcea565b3fbc698bb19cdf6deb8500000000000000000000000000000000000000000000000000000000000000010000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000',
  gas: '',
  maxFeePerGas: '0x7aef40a00',
  maxPriorityFeePerGas: '0x7aef40a00',
  nonce: '0xb5e',
};
