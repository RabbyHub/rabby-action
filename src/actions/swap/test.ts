import {
  apiProvider,
  ETH_CHAIN_ID,
  formatProvider,
  SENDER,
  walletProvider,
} from '../../../__mocks__';
import { ParsedTransactionActionData } from '../../types';
import { fetchDataSwap } from './fetchData';
import { formatSecurityEngineSwap } from './formatSecurityEngine';
import { parseTxData, preExecData, txData } from './mocks';
import { parseActionSwap } from './parseAction';
import {
  parseAction,
  fetchActionRequiredData,
  formatSecurityEngineContext,
} from '../..';

/**
 * Rabby Swap
 */
test.each([
  [parseActionSwap, fetchDataSwap, formatSecurityEngineSwap],
  [parseAction, fetchActionRequiredData, formatSecurityEngineContext],
])('Swap', async (_parseAction, _fetchData, _format) => {
  const actionData = _parseAction({
    type: 'transaction',
    data: parseTxData['action'],
    balanceChange: preExecData.balance_change,
    sender: txData.from,
    preExecVersion: preExecData.pre_exec_version,
    gasUsed: preExecData.gas.gas_used,
    tx: txData,
  }) as ParsedTransactionActionData;
  expect(actionData).toMatchSnapshot('parseActionSwap');

  const requireData = await _fetchData({
    type: 'transaction',
    actionData,
    contractCall: parseTxData.contract_call,
    chainId: ETH_CHAIN_ID,
    sender: SENDER,
    walletProvider,
    tx: txData,
    apiProvider,
  });

  expect(requireData).toMatchSnapshot('fetchDataSwap');

  const ctx = await _format({
    type: 'transaction',
    actionData,
    requireData,
    chainId: ETH_CHAIN_ID,
    isTestnet: false,
    provider: formatProvider,
  });
  expect(ctx).toMatchSnapshot('formatSecurityEngineSwap');
});
