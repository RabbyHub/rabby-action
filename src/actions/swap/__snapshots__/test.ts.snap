// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Swap: fetchDataSwap 1`] = `
{
  "bornAt": 1591388241,
  "extraState": {
    "hasInteraction": [Function],
  },
  "hasInteraction": true,
  "id": "0x6352a56caadc4f1e25cd6c75970fa768a3304e64",
  "protocol": {
    "id": "uniswap2",
    "logo_url": "https://static.debank.com/image/project/logo_url/uniswap2/87a541b3b83b041c8d12119e5a0d19f0.png",
    "name": "Uniswap V2",
  },
  "rank": 65,
  "receiverInWallet": true,
  "sender": "0x5853ed4f26a3fcea565b3fbc698bb19cdf6deb85",
}
`;

exports[`Swap: fetchDataSwap 2`] = `
{
  "bornAt": 1591388241,
  "extraState": {
    "hasInteraction": [Function],
  },
  "hasInteraction": true,
  "id": "0x6352a56caadc4f1e25cd6c75970fa768a3304e64",
  "protocol": {
    "id": "uniswap2",
    "logo_url": "https://static.debank.com/image/project/logo_url/uniswap2/87a541b3b83b041c8d12119e5a0d19f0.png",
    "name": "Uniswap V2",
  },
  "rank": 65,
  "receiverInWallet": true,
  "sender": "0x5853ed4f26a3fcea565b3fbc698bb19cdf6deb85",
}
`;

exports[`Swap: formatSecurityEngineSwap 1`] = `
{
  "swap": {
    "chainId": "eth",
    "contractAddress": "0x6352a56caadc4f1e25cd6c75970fa768a3304e64",
    "from": "0x5853ed4f26a3fcea565b3fbc698bb19cdf6deb85",
    "id": "0x6352a56caadc4f1e25cd6c75970fa768a3304e64",
    "receiveTokenIsFake": false,
    "receiveTokenIsScam": false,
    "receiver": "0x5853ed4f26a3fcea565b3fbc698bb19cdf6deb85",
    "receiverInWallet": true,
    "slippageTolerance": 0.5,
    "usdValuePercentage": -0.0063510049707952015,
  },
}
`;

exports[`Swap: formatSecurityEngineSwap 2`] = `
{
  "swap": {
    "chainId": "eth",
    "contractAddress": "0x6352a56caadc4f1e25cd6c75970fa768a3304e64",
    "from": "0x5853ed4f26a3fcea565b3fbc698bb19cdf6deb85",
    "id": "0x6352a56caadc4f1e25cd6c75970fa768a3304e64",
    "receiveTokenIsFake": false,
    "receiveTokenIsScam": false,
    "receiver": "0x5853ed4f26a3fcea565b3fbc698bb19cdf6deb85",
    "receiverInWallet": true,
    "slippageTolerance": 0.5,
    "usdValuePercentage": -0.0063510049707952015,
  },
}
`;

exports[`Swap: parseActionSwap 1`] = `
{
  "swap": {
    "balanceChange": {
      "success": true,
      "support": true,
    },
    "minReceive": {
      "amount": 0.00005,
      "chain": "matic",
      "credit_score": 4530369.670714375,
      "decimals": 6,
      "display_symbol": null,
      "id": "0xc2132d05d31c914a87c6611c10748aeb04b58e8f",
      "is_core": true,
      "is_scam": false,
      "is_suspicious": false,
      "is_verified": true,
      "is_wallet": true,
      "logo_url": "https://static.debank.com/image/eth_token/logo_url/0xdac17f958d2ee523a2206206994597c13d831ec7/464c0de678334b8fe87327e527bc476d.png",
      "min_amount": 0.00005,
      "min_raw_amount": "50",
      "name": "(PoS) Tether USD",
      "optimized_symbol": "USDT",
      "price": 1.00038,
      "price_24h_change": 0.000009996301368559145,
      "protocol_id": "",
      "symbol": "USDT",
      "time_at": 1599512847,
    },
    "payToken": {
      "amount": 0.000192647154008608,
      "chain": "matic",
      "credit_score": 100000000,
      "decimals": 18,
      "display_symbol": null,
      "id": "matic",
      "is_core": true,
      "is_scam": false,
      "is_suspicious": false,
      "is_verified": true,
      "is_wallet": true,
      "logo_url": "https://static.debank.com/image/matic_token/logo_url/matic/6f5a6b6f0732a7a235131bd7804d357c.png",
      "name": "MATIC",
      "optimized_symbol": "MATIC",
      "price": 0.5226,
      "price_24h_change": -0.08716157205240184,
      "protocol_id": "",
      "raw_amount": "192647154008608",
      "symbol": "MATIC",
      "time_at": null,
    },
    "receiveToken": {
      "amount": 0.0001,
      "chain": "matic",
      "credit_score": 4530369.670714375,
      "decimals": 6,
      "display_symbol": null,
      "id": "0xc2132d05d31c914a87c6611c10748aeb04b58e8f",
      "is_core": true,
      "is_scam": false,
      "is_suspicious": false,
      "is_verified": true,
      "is_wallet": true,
      "logo_url": "https://static.debank.com/image/eth_token/logo_url/0xdac17f958d2ee523a2206206994597c13d831ec7/464c0de678334b8fe87327e527bc476d.png",
      "min_amount": 0.00005,
      "min_raw_amount": "50",
      "name": "(PoS) Tether USD",
      "optimized_symbol": "USDT",
      "price": 1.00038,
      "price_24h_change": 0.000009996301368559145,
      "protocol_id": "",
      "symbol": "USDT",
      "time_at": 1599512847,
    },
    "receiver": "0x5853ed4f26a3fcea565b3fbc698bb19cdf6deb85",
    "slippageTolerance": 0.5,
    "usdValueDiff": "-0.0000006394026848985408",
    "usdValuePercentage": -0.0063510049707952015,
  },
}
`;

exports[`Swap: parseActionSwap 2`] = `
{
  "swap": {
    "balanceChange": {
      "success": true,
      "support": true,
    },
    "minReceive": {
      "amount": 0.00005,
      "chain": "matic",
      "credit_score": 4530369.670714375,
      "decimals": 6,
      "display_symbol": null,
      "id": "0xc2132d05d31c914a87c6611c10748aeb04b58e8f",
      "is_core": true,
      "is_scam": false,
      "is_suspicious": false,
      "is_verified": true,
      "is_wallet": true,
      "logo_url": "https://static.debank.com/image/eth_token/logo_url/0xdac17f958d2ee523a2206206994597c13d831ec7/464c0de678334b8fe87327e527bc476d.png",
      "min_amount": 0.00005,
      "min_raw_amount": "50",
      "name": "(PoS) Tether USD",
      "optimized_symbol": "USDT",
      "price": 1.00038,
      "price_24h_change": 0.000009996301368559145,
      "protocol_id": "",
      "symbol": "USDT",
      "time_at": 1599512847,
    },
    "payToken": {
      "amount": 0.000192647154008608,
      "chain": "matic",
      "credit_score": 100000000,
      "decimals": 18,
      "display_symbol": null,
      "id": "matic",
      "is_core": true,
      "is_scam": false,
      "is_suspicious": false,
      "is_verified": true,
      "is_wallet": true,
      "logo_url": "https://static.debank.com/image/matic_token/logo_url/matic/6f5a6b6f0732a7a235131bd7804d357c.png",
      "name": "MATIC",
      "optimized_symbol": "MATIC",
      "price": 0.5226,
      "price_24h_change": -0.08716157205240184,
      "protocol_id": "",
      "raw_amount": "192647154008608",
      "symbol": "MATIC",
      "time_at": null,
    },
    "receiveToken": {
      "amount": 0.0001,
      "chain": "matic",
      "credit_score": 4530369.670714375,
      "decimals": 6,
      "display_symbol": null,
      "id": "0xc2132d05d31c914a87c6611c10748aeb04b58e8f",
      "is_core": true,
      "is_scam": false,
      "is_suspicious": false,
      "is_verified": true,
      "is_wallet": true,
      "logo_url": "https://static.debank.com/image/eth_token/logo_url/0xdac17f958d2ee523a2206206994597c13d831ec7/464c0de678334b8fe87327e527bc476d.png",
      "min_amount": 0.00005,
      "min_raw_amount": "50",
      "name": "(PoS) Tether USD",
      "optimized_symbol": "USDT",
      "price": 1.00038,
      "price_24h_change": 0.000009996301368559145,
      "protocol_id": "",
      "symbol": "USDT",
      "time_at": 1599512847,
    },
    "receiver": "0x5853ed4f26a3fcea565b3fbc698bb19cdf6deb85",
    "slippageTolerance": 0.5,
    "usdValueDiff": "-0.0000006394026848985408",
    "usdValuePercentage": -0.0063510049707952015,
  },
}
`;
